clang/lib/Frontend/CompileJobCacheKey.cpp

```diff
@@ -110,6 +107,51 @@ clang::createCompileJobCacheKey(ObjectStore &CAS, DiagnosticsEngine &Diags,
   return createCompileJobCacheKeyForArgs(CAS, Argv, *RootID, IsIncludeTree);
 }
 
+static void canonicalizeForCaching(llvm::cas::ObjectStore &CAS,
+                                   DiagnosticsEngine &Diags,
+                                   CompilerInvocation &Invocation,
+                                   bool &WriteOutputAsCASID) {
```

Comment:
How about instead of passing `bool &WriteOutputAsCASID` as parameter you introduce a struct like
```
struct CompileJobCachingOptions {
  bool WriteOutputAsCASID;
};
```
returned via
```
CompileJobCachingOptions canonicalizeForCaching(llvm::cas::ObjectStore &CAS, ...) {
```
?

That way we can conveniently add more extracted options, like the `DisableCachedCompileJobReplay` from https://github.com/apple/llvm-project/pull/5300.

Reply:
Good idea, done in latest commit.  I also rebased and picked up `DisableCachedCompileJobReplay` in the new struct.

I kept it as an out-parameter instead of a return value because I still want to do canonicalization and getting the cache key in one function.